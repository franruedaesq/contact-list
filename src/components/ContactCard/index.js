import React, { useState, useEffect } from 'react';
import './index.scss';
import { setSelectedContact } from '../../redux/contacts/contactsSlice';
import { useDispatch } from 'react-redux';

export default function ContactCard({ data, activeId }) {
  const {
    name,
    lastName,
    email,
    phone,
    street,
    city,
    state,
    postCode,
    steetNumber,
    img,
    id,
  } = data;
  const dispatch = useDispatch();
  const [showCard, setShowCard] = useState(false);

  const handleClick = () => {
    dispatch(setSelectedContact(id));
    console.log(showCard);
    if (!showCard) {
      setShowCard(!showCard);
    }
  };

  const closeCard = () => {
    dispatch(setSelectedContact(''));
    setShowCard(false);
  };

  useEffect(() => {
    setShowCard(activeId === id);
  }, [id, activeId]);

  return (
    <article
      className={`contactCard ${showCard ? 'contactCard--show' : ''}`}
      onClick={handleClick}
    >
      <h3
        className={`contactCard_title contactCard_title${
          showCard ? '--show' : ''
        }`}
      >
        {name} {lastName}
      </h3>
      <div
        className={`contactCard_body ${
          showCard ? 'contactCard_body--show' : 'contactCard_body--hide'
        }`}
      >
        <picture className='contactCard_img'>
          <source srcset={img} media='(min-width: 800px)' />
          <img src={img} alt={`${name} ${lastName}`} />
        </picture>
        <div>
          <span className='contactCard_item'>
            <strong>E-mail:</strong>
            {email}
          </span>
          <span className='contactCard_item'>
            <strong>Phone:</strong>
            {phone}
          </span>
          <span className='contactCard_item'>
            <strong>Street:</strong>
            {steetNumber} {street}
          </span>
          <span className='contactCard_item'>
            <strong>City:</strong>
            {city}
          </span>
          <span className='contactCard_item'>
            <strong>State:</strong>
            {state}
          </span>
          <span className='contactCard_item'>
            <strong>Postcode:</strong>
            {postCode}
          </span>
        </div>
      </div>
      <button
        className={`contactCard_close ${
          showCard ? 'contactCard_close--show' : ''
        }`}
        aria-label='Close Contact Card'
        onClick={closeCard}
      >
        &times;
      </button>
    </article>
  );
}
