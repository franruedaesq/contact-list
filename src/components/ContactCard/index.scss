.contactCard {
  position: relative;
  display: flex;
  flex-direction: column;
  background-color: #ffffff;
  border-radius: 7px;
  padding: 5px;
  padding-left: 15px;
  margin: 10px 5px;
  width: 470px;
  height: 40px;
  align-items: baseline;
  box-shadow: 0px 4px 10px #c7c7c7;
  cursor: pointer;
  transition: all 1s;

  @media (max-width: 500px) {
    width: auto;
  }

  &--show {
    height: 250px;
  }

  &_header {
    & h3 {
      display: none;
    }
    @media (max-width: 500px) {
      display: flex;
      height: 40px;
      justify-content: space-between;
      align-items: center;
      & h3 {
        margin: 0;
        display: inline;
      }
    }
  }

  &_img {
    min-width: 130px;
    border-radius: 30%;
    align-self: center;
    overflow: hidden;
    margin: 5px;
    display: flex;
    align-items: center;
    justify-content: center;

    & img {
      width: 100%;
    }

    @media (max-width: 500px) {
      min-width: auto;
      width: 50px;
      min-height: 60px;
      margin-right: 30px;
      border-radius: 50%;
      margin-right: 40px;
    }
  }
  &_title {
    font-size: 1rem;
    font-weight: 700;
    margin: 7px 0 5px 0;
    &--hide {
      font-size: 1rem;
      font-weight: normal;
    }

    @media (max-width: 500px) {
      display: none;
    }
  }
  &_body {
    display: flex;
    width: 100%;
    overflow: hidden;

    @media (max-width: 500px) {
      flex-direction: column;
    }
  }
  &_item {
    margin: 3px;
    display: flex;
    font-size: 1rem;

    & strong {
      width: 75px;
      text-align: right;
      margin-right: 10px;
    }

    @media (max-width: 500px) {
      flex-direction: column;

      & strong {
        text-align: left;
      }
    }
  }

  &_close {
    border: none;
    position: absolute;
    top: 5px;
    right: 5px;
    background-color: transparent;
    cursor: pointer;
    visibility: hidden;
    font-size: 20px;
    &--show {
      visibility: visible;
      top: 10px;
    }
  }
}
