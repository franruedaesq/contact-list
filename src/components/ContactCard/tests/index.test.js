import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import { render, fireEvent } from '@testing-library/react';
import ContactCard from '../index';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { hideActiveId, props, showActiveId } from './helper';

describe('test', () => {
  const initialState = {
    contactList: {},
    contactSelected: '',
  };
  const mockStore = configureStore();
  let store;
  store = mockStore(initialState);

  it('should match snapshot', () => {
    const { data, index } = props;
    const activeId = showActiveId;

    const component = render(
      <Provider store={store}>
        <ContactCard data={data} activeId={activeId} key={index} />
      </Provider>
    );
    const { asFragment } = component;
    expect(asFragment(component)).toMatchSnapshot();
  });
  it('should render a component, component should not be null', () => {
    const { data, index } = props;
    const activeId = showActiveId;

    const component = render(
      <Provider store={store}>
        <ContactCard data={data} activeId={activeId} key={index} />
      </Provider>
    );
    expect(component).not.toBeNull();
  });
  it('should be hide', () => {
    const { data, index } = props;
    const activeId = hideActiveId;
    const component = render(
      <Provider store={store}>
        <ContactCard data={data} activeId={activeId} key={index} />
      </Provider>
    );
    const card = component.getByTestId('contactCard_container');
    expect(card).toHaveClass('contactCard--hide');
  });
  it('should be show', () => {
    const { data, index } = props;
    const activeId = showActiveId;
    const component = render(
      <Provider store={store}>
        <ContactCard data={data} activeId={activeId} key={index} />
      </Provider>
    );
    const card = component.getByTestId('contactCard_container');
    expect(card).toHaveClass('contactCard--show');
  });
  it('should hide data when close button is clicked', () => {
    const { data, index } = props;
    const activeId = showActiveId;
    const component = render(
      <Provider store={store}>
        <ContactCard data={data} activeId={activeId} key={index} />
      </Provider>
    );
    const closeButton = component.getByTestId('contactCard_close');
    fireEvent.click(closeButton);
    const card = component.getByTestId('contactCard_container');
    expect(card).toHaveClass('contactCard--hide');
  });
  it('should show data when the card is clicked', () => {
    const { data, index } = props;
    const activeId = hideActiveId;
    const component = render(
      <Provider store={store}>
        <ContactCard data={data} activeId={activeId} key={index} />
      </Provider>
    );
    const card = component.getByTestId('contactCard_container');
    fireEvent.click(card);
    expect(card).toHaveClass('contactCard--show');
  });
});
