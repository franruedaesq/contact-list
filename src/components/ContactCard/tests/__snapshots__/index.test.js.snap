// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test should match snapshot 1`] = `
<DocumentFragment>
  <article
    class="contactCard contactCard--show"
    data-testid="contactCard_container"
  >
    <h3
      class="contactCard_title contactCard_title--show"
    >
      Monica Lewis
    </h3>
    <div
      class="contactCard_body contactCard_body--show"
    >
      <div
        class="contactCard_header"
      >
        <h3>
          Monica Lewis
        </h3>
        <picture
          class="contactCard_img"
        >
          <source
            media="(min-width: 800px)"
            srcset="http://www.img.com"
          />
          <img
            alt="Monica Lewis"
            src="http://www.img.com"
          />
        </picture>
      </div>
      <div>
        <span
          class="contactCard_item"
        >
          <strong>
            E-mail:
          </strong>
          monicalewis@email.com
        </span>
        <span
          class="contactCard_item"
        >
          <strong>
            Phone:
          </strong>
          254233343
        </span>
        <span
          class="contactCard_item"
        >
          <strong>
            Street:
          </strong>
          2500 Hardly Steet
        </span>
        <span
          class="contactCard_item"
        >
          <strong>
            City:
          </strong>
          Rosario
        </span>
        <span
          class="contactCard_item"
        >
          <strong>
            State:
          </strong>
          Santa Fe
        </span>
        <span
          class="contactCard_item"
        >
          <strong>
            Postcode:
          </strong>
          5300
        </span>
      </div>
    </div>
    <button
      aria-label="Close Contact Card"
      class="contactCard_close contactCard_close--show"
      data-testid="contactCard_close"
    >
      Ã—
    </button>
  </article>
</DocumentFragment>
`;
